File: Лекции/p02_Binary_Phase_Grating_1D.m

%% 1D Rectangular Phase Grating
clc
% close all
clear all

%% Defining Grating Parameters
N = 500; % Define Matrix size
A = zeros(1,N); % Define a row Matrix by assigning 0 to all pixels
P = 50; % Define the period of the grating
FF = 0.75; % Define fill factor

%% Constructing the Grating
% V.1
% for q = 1:N;
%     if rem(q,P) < P*FF; %Use remainder function ‘rem’ to construct
%                       %the grating
%        A(1,q) = 1;
%     end
% end
%     A = repmat(A,N,1); %replicate the row to create a 2D grating

% V.2 (Alternative code for constructing the grating)
O = ones(N,round(FF*P));
Z = zeros(N, P-round(FF*P));
unit = [O Z];
A = repmat(unit,1,N/P); % Replicate to create a 1D grating

figure
plot(A(1,:));

%% Creation of a phase grating
B = exp(1i.*A');

figure;
plot(angle(B(1,:)));

figure;
imshow(mat2gray(angle(B)));

%% Observing the grating output in the far field
E = fftshift(fft(B)); % fftshift is used to re-order the terms in their natural order
IN = (abs(E)/(N)).*(abs(E)/(N)); % Calculating intensity

figure;
colormap(gray); %colormap(gray) is used to display grayscale
imagesc(A);% imagesc is used to display a high constrast image

figure;
colormap(gray);
imagesc(IN);

figure
plot(IN(:,round(N/2) + 1)); % round(N/2) + 

File: opticalCalculation/p02_Binary_Phase_Grating_1D.py
import numpy as np
import matplotlib.pyplot as plt

def create_rectangular_phase_grating(N=500, P=50, FF=0.75):
    """
    Рассчитывает 1D прямоугольную фазовую решетку и отображает выход решетки в дальней зоне.

    Parameters:
    N (int): Размер матрицы
    P (int): Период решетки
    FF (float): Скважность (Fill factor)

    Returns:
    None
    """
    # Определение параметров решетки
    A = np.zeros(N)  # Определяем строковую матрицу, присваивая 0 всем пикселям

    # Конструирование решетки
    O = np.ones((N, int(round(FF * P))))
    Z = np.zeros((N, P - int(round(FF * P))))
    unit = np.hstack((O, Z))
    A = np.tile(unit, (1, N // P))

    # Построение графика
    plt.figure()
    plt.plot(A[0, :])
    plt.show()

    # Создание фазовой решетки
    B = np.exp(1j * A.T)

    # Построение угловой части B
    plt.figure()
    plt.plot(np.angle(B[0, :]))
    plt.show()

    # Отображение матрицы
    plt.figure()
    plt.imshow(np.angle(B), cmap='gray')
    plt.show()

    # Наблюдение выхода решетки в дальней зоне
    E = np.fft.fftshift(np.fft.fft(B))  # fftshift используется для упорядочивания членов в их естественном порядке
    IN = (np.abs(E) / N) ** 2  # Вычисление интенсивности

    # Отображение изображения с высоким контрастом
    plt.figure()
    plt.imshow(A, cmap='gray')
    plt.show()

    # Отображение интенсивности
    plt.figure()
    plt.imshow(IN, cmap='gray')
    plt.show()

    # Построение интенсивности по центру
    plt.figure()
    plt.plot(IN[:, N // 2])
    plt.show()
